### ISSUE TITLE
conv2d support for bfloat16

### ISSUE THREAD

**Issue by geandwor** [tensorflow/tensorflow#21612]
As in the reference for r1.8.0 of tensorflow, the tf.nn.conv2d support bfloat16, but actually when run conv2d with bfloat16 format, got error:
No OpKernel was registered to support Op 'Conv2D' with these attrs.  Registered devices: [CPU,GPU], Registered kernels:
  device='GPU'; T in [DT_DOUBLE]
  device='GPU'; T in [DT_FLOAT]
  device='GPU'; T in [DT_HALF]
  device='CPU'; T in [DT_DOUBLE]
  device='CPU'; T in [DT_FLOAT]
  device='CPU'; T in [DT_HALF]

So is reference wrong, or it is something which would be added later?

**Comment by sabhiram** [comment]
The `DataType` definition calls out that 

```
DT_BFLOAT16 = 14;  // Float32 truncated to 16 bits.  Only for cast ops.
```

You could use an artificially less precise Float32 to emulate 16 bits for the time being.

**Comment by geandwor** [comment]
thanks.  so reference of conv2d is misleading.





On 2018-08-14 22:00 , Shaba Abhiram<mailto:notifications@github.com> Wrote:


The DataType definition calls out that

DT_BFLOAT16 = 14;  // Float32 truncated to 16 bits.  Only for cast ops.


You could use an artificially less precise Float32 to emulate 16 bits for the time being.

**Comment by michaelisard** [comment]
It looks as if the docs are wrong yes.

**Comment by yongtang** [comment]
It looks like the documentation of `conv2d` is automatically generated from tensorflow/core/api_def/base_api/api_def_Conv2D.pbtxt, and `nn_ops.cc`:
https://github.com/tensorflow/tensorflow/blob/62191da0819b25906c1b2ed96159cfe36ba00383/tensorflow/core/ops/nn_ops.cc#L265-L270

Deleting `bfloat16` from `nn_ops.cc` will fix the doc issue although that may break API backward compatibility I think.

Alternatively it might possible to create a python wrapper of `def conv2d` in `nn_ops.py`, that calls `gen_nn_ops.conv2d`. Then inside `def conv2d` we could have the correct documentation.

Not sure the best way to handle this situation. Would be happy to create a PR with some guidance.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @MarkDaoust: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
